python/cudf/cudf/core/join/_join_helpers.py

```diff
@@ -0,0 +1,196 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.
+from __future__ import annotations
+
+import warnings
+from typing import TYPE_CHECKING, Any, Iterable
+
+import numpy as np
+import pandas as pd
+
+import cudf
+from cudf.core.dtypes import CategoricalDtype
+
+if TYPE_CHECKING:
+    from cudf._typing import Dtype
+    from cudf.core.column import ColumnBase
+    from cudf.core.frame import Frame
+
+
+class _Indexer:
+    # Indexer into a column (either a data column or index level).
+    #
+    # >>> df
+    #    a
+    # b
+    # 4  1
+    # 5  2
+    # 6  3
+    # >>> _Indexer("a", column=True).get(df)  # returns column "a" of df
+    # >>> _Indexer("b", index=True).get(df)  # returns index level "b" of df
+
+    def __init__(self, name: Any, column=False, index=False):
+        if column and index:
+            raise ValueError("Cannot specify both column and index")
+        self.name = name
+        self.column, self.index = column, index
+
+    def get(self, obj: Frame) -> ColumnBase:
+        # get the column from `obj`
+        if self.column:
+            return obj._data[self.name]
+        else:
+            if obj._index is not None:
+                return obj._index._data[self.name]
+        raise KeyError()
+
+    def set(self, obj: Frame, value: ColumnBase):
+        # set the colum in `obj`
+        if self.column:
+            obj._data[self.name] = value
+        else:
+            if obj._index is not None:
+                obj._index._data[self.name] = value
+            else:
+                raise KeyError()
+
+
+def _frame_select_by_indexers(
+    frame: Frame, indexers: Iterable[_Indexer]
+) -> Frame:
+    # Select columns from the given `Frame` using `indexers`,
+    # and return a new `Frame`.
+    index_data = frame._data.__class__()
+    data = frame._data.__class__()
+
+    for idx in indexers:
+        if idx.index:
+            index_data[idx.name] = idx.get(frame)
+        else:
+            data[idx.name] = idx.get(frame)
+
+    result_index = cudf.Index._from_data(index_data) if index_data else None
+    result = cudf.core.frame.Frame(data=data, index=result_index)
+    return result
+
+
+def _match_join_keys(lcol: ColumnBase, rcol: ColumnBase, how: str) -> Dtype:
+    # cast the keys lcol and rcol to a common dtype
+
+    ltype = lcol.dtype
+    rtype = rcol.dtype
+
+    # if either side is categorical, different logic
+    if isinstance(ltype, CategoricalDtype) or isinstance(
+        rtype, CategoricalDtype
+    ):
+        return _match_categorical_dtypes(ltype, rtype, how)
+
+    if pd.api.types.is_dtype_equal(ltype, rtype):
+        return ltype
+
+    if (np.issubdtype(ltype, np.number)) and (np.issubdtype(rtype, np.number)):
+        common_type = (
+            max(ltype, rtype)
+            if ltype.kind == rtype.kind
+            else np.find_common_type([], (ltype, rtype))
+        )
+
+    elif np.issubdtype(ltype, np.datetime64) and np.issubdtype(
+        rtype, np.datetime64
```

Comment:
Do we handle timedeltas? Should we handle timedeltas?

Reply:
We don't -- do you think I can naively copy the logic we have here for datetimes? If so, happy to tack it on. If not, we could consider tackling this in another PR?

Reply:
Did the previous implementation handle timedeltas? If it did we should make sure to cover them in this PR as well. If not, we can do it as a follow up, but we should ultimately handle them.

Reply:
We did not handle timedeltas before. We can here, it's the same logic essentially as that of DateTime, but then it'd require tests and inflate things.

Reply:
Would be nice, but happy to do it as a follow up as well. Maybe in the same PR as decimal?